### ISSUE TITLE
Invalid memory access when extending a BAT during appends

### ISSUE THREAD

**Issue by swingbit** [MonetDB/MonetDB#7225]
**Describe the bug**
When an append goes over the initial capacity and the BAT needs to be extended, invalid memory access occurs.

**To Reproduce**
Take this minimal UDF, which simply creates a BAT with a number of strings:
```
str
bat_HeapExtendTest(bat *r) {
	BAT *bn = NULL;
	char buf[20];
	BUN initial_capacity = 0; // This actually creates a BAT with initial capacity = 256
	BUN count = 257; // Just one over the initial capacity, will trigger a BATextend()

	if ((bn = COLnew(0, TYPE_str, initial_capacity, TRANSIENT)) == NULL) goto fail;

	for (BUN i = 1; i <= count; i++) {
		sprintf(buf, "String " BUNFMT, i);
		if (bunfastapp(bn, buf) != GDK_SUCCEED) goto fail;
	}

	BBPkeepref(*r = bn->batCacheid);

	return MAL_SUCCEED;

fail:
	if (bn)
		BBPreclaim(bn);
	throw(MAL, "heapExtendTest", GDK_EXCEPTION);
}

```
And map it to SQL:
```
command("sys", "heapExtendTest", bat_HeapExtendTest, false, "", args(1, 1, batarg("", str))),
```
```
CREATE OR REPLACE FUNCTION heapExtendTest() RETURNS TABLE(s string) EXTERNAL NAME sys."heapExtendTest";
```

Finally call it:
```
select * from heapExtendTest();
```

This produces undefined behaviour (mostly SIGSEGV), and when compiled in DEBUG mode:
```
<path>/src/gdk/gdk_utils.c:1987: GDKfree: Assertion `((char *) s)[i] == '\xBD'' failed.
```
On a different UDF, which boils down to the same issue, I got the same assertion, but really during the `GDKrealloc()` that is triggered by `HEAPextend()`. Both cases are symptoms of invalid memory access.

**IMPORTANT**. This UDF works fine when `initial_capacity >= count`, i.e. if the BAT capacity doesn't need to be extended. I suppose this was not caught before because most internal functions do know the required result capacity in advance.

**Expected behavior**
A table with strings from `String 1` to `String 257`.

**Software versions**
 - MonetDB version number 1.41.14 (hg id: e465e28319)
 - OS and version: Fedora 35
 - self-installed and compiled

**Comment by swingbit** [comment]
I might be wrong, but I believe the issue is in `tfastins`:
```
static inline gdk_return __attribute__((__warn_unused_result__))
tfastins(BAT *b, BUN p, const void *v)
{
	if (p > BATcapacity(b)) {
		if (p >= BUN_MAX) {
			GDKerror("tfastins: too many elements to accommodate (" BUNFMT ")\n", BUN_MAX);
			return GDK_FAIL;
		}
		BUN sz = BATgrows(b);
		if (sz <= p)
			sz = p + BATTINY;
		gdk_return rc = BATextend(b, sz);
		if (rc != GDK_SUCCEED)
			return rc;
	}
	return tfastins_nocheck(b, p, v);
}
```

When `bunfastapp` calls `tfastins`, `p = BATcount(b)`.
This means that when we are about to insert tuple 257, `p = 256`.
Then the condition in `tfastins` reads:
```
if (256 > 256) {
```
Therefore it does not extend it, though the capacity has been used already.
Tuple 257 is written on unallocated memory, which sometimes works, sometimes fails.
If it doesn't fail at this round, next one will satisfy the condition (too late), and the `realloc()` will fail, because it finds unallocated memory that has been wrongly overwritten.

If I'm not mistaken, it's an off by 1 error, and the fix should just be:
```
if (p >= BATcapacity(b)) {
```

Oct2021 actually works this way.

**Comment by sjoerdmullender** [comment]
Should be fixed with hg changeset ce2937077a3d.
